src/thor/triplegbuilder.cc

```diff
@@ -92,25 +92,47 @@ void AssignAdmins(const AttributesController& controller,
 void SetShapeAttributes(const AttributesController& controller,
                         const GraphTile* tile,
                         const DirectedEdge* edge,
-                        const std::shared_ptr<sif::DynamicCost>& costing,
-                        std::vector<PointLL>::const_iterator shape_begin,
-                        std::vector<PointLL>::const_iterator shape_end,
+                        std::vector<PointLL>& shape,
+                        size_t shape_begin,
                         TripLeg& trip_path,
-                        uint32_t second_of_week,
-                        double edge_percentage) {
+                        double src_pct,
+                        double tgt_pct,
+                        double edge_seconds,
+                        bool cut_for_traffic) {
+  // TODO: if this is a transit edge then the costing will throw
+
   if (trip_path.has_shape_attributes()) {
-    // calculates total edge time and total edge length
-    // TODO: you can get this directly from the path edge by taking its cost and subtracting off
-    // the transition cost that it also now contains
-    double edge_time =
-        costing->EdgeCost(edge, tile, second_of_week).secs * edge_percentage; // seconds
-    // TODO: get the measured length from shape (full shape) to increase precision
-    double edge_length = edge->length() * edge_percentage; // meters
+    // A list of percent along the edge and corresponding speed (meters per second)
+    std::vector<std::tuple<double, double>> speeds;
+    double speed = (edge->length() * (tgt_pct - src_pct)) / edge_seconds;
+    speeds.emplace_back(tgt_pct, speed);
+    if (cut_for_traffic) {
+      // TODO: we'd like to use the speed from traffic here but because there are synchronization
+      // problems with those records changing between when we used them to make the path and when we
+      // try to grab them again here, we instead rely on the total time from PathInfo and just do the
+      // cutting for now
+      const auto& traffic_speed = tile->trafficspeed(edge);
+      if (traffic_speed.valid()) {
+        speeds.clear();
+        speeds.emplace_back(traffic_speed.breakpoint1 / 255.0, speed);
+        if (traffic_speed.breakpoint1 < 255) {
+          speeds.emplace_back(traffic_speed.breakpoint2 / 255.0, speed);
```

Comment:
Still trying to trace how this is used later in the function, but first thought is that this would emplace to `speeds` even if `breakpoint2` is 0 and live speed is unknown. Not sure yet if this is intended.

Reply:
yeah i was going back and forth on this. this needs a change. thanks!